routing/matcher.go

```diff
@@ -139,6 +133,75 @@ func canonicalizeHeaderRegexps(hrx map[string][]*regexp.Regexp) map[string][]*re
 	return chrx
 }
 
+// extracts the expected wildcard param names and returns them in reverse order
+func extractWildcardParamNames(r *Route) []string {
+	path := r.path
+	if path == "" {
+		path = r.pathSubtree
+	}
+	path = httppath.Clean(path)
+
+	var wildcards []string
+	pathTokens := strings.Split(path, "/")
+	for _, token := range pathTokens {
+		if len(token) > 1 && (token[0] == ':' || token[0] == '*') {
+			//prepend
+			wildcards = append([]string{token[1:]}, wildcards...)
+		}
+	}
+
+	if strings.HasSuffix(path, "/*") ||
+		r.path == "" && r.pathSubtree != "" && !freeWildcardRx.MatchString(path) {
+
+		wildcards = append([]string{"*"}, wildcards...)
+	}
+
+	return wildcards
+}
+
+func hasFreeWildcardParam(r *Route) bool {
+	return r.pathSubtree != "" || freeWildcardRx.MatchString(httppath.Clean(r.path))
+}
+
+// returns a cleaned path where all wildcard names have been replaced with *
+func normalizePath(r *Route) string {
+	path := r.path
+	if path == "" {
+		path = r.pathSubtree
+	}
+	path = httppath.Clean(path)
+
+	var sb strings.Builder
+	for i := 0; i < len(path); i++ {
+		c := path[i]
+		if c == '/' {
+			sb.WriteByte(path[i])
+		} else {
+			nextSlash := strings.IndexByte(path[i:], '/')
+			nextSlashExists := true
+			if nextSlash == -1 {
+				nextSlash = len(path)
+				nextSlashExists = false
+			} else {
+				nextSlash += i
+			}
+			if c == ':' || c == '*' {
+				if nextSlashExists {
+					sb.WriteByte(':')
```

Comment:
will it accept paths like `/foo/*bar/baz` this way? I think if nextSlashExists is true, and `c == '*'` then it should return an error.

Reply:
I thought that `:` and `*` is interchangeable for non-free wildcard params. It's worth confirming this was the previous behavior.

Reply:
I pushed a change to match the previous behavior where only free wildcard params are prefixed by `*` and are only last.